### PR TITLE
shell: Reimplement shell.plot with the grid.

### PR DESCRIPTION
**Pull request by mvollmer**

### CODE REVIEW

**Comment by mvollmer** [comment]
This also fixes some long standing bugs re archives, most importantly "bridge: Always flush metrics data before sending a meta message."

**Comment by mvollmer** [comment]
There are some JavaScript exceptions.

**Comment by mvollmer** [comment]
The cpu plot scaling is all wrong.

File: pkg/base1/cockpit.js [hunk 1]
```diff
@@ -2763,19 +2763,22 @@ function full_scope(cockpit, $, po) {
             throw "invalid date or offset";
     }
 
-    function MetricsChannel(interval, options) {
+    function MetricsChannel(interval, orig_options) {
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: I would prefer if fallbacks were implemented as ```options, ...```. In other words you specify N dicts of options, and we fallback to the next one if the first one is ```not-supported```

**Comment by mvollmer** [comment]
Scrolling should not create a new grid.

File: pkg/base1/cockpit.js [hunk 2]
```diff
@@ -2785,7 +2788,10 @@ function full_scope(cockpit, $, po) {
         self.series = cockpit.series(interval, options.cache);
         delete options.cache;
 
-        function transfer(options, callback) {
+        function have_archives() {
+        }
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Unused?

File: pkg/base1/cockpit.js [hunk 3]
```diff
@@ -2794,9 +2800,21 @@ function full_scope(cockpit, $, po) {
...
+                        (close_options.problem == "not-supported" || close_options.problem == "not-found")) {
+                        options.source = orig_options.fallback_source;
+                        options.metrics = orig_options.fallback_metrics;
+                        transfer(options, callback, true);
+                    } else if (close_options.problem) {
+                        console.warn("metrics channel failed: " + close_options.problem);
+                    } else if (options.source == "pcp-archive") {
+                        if (!self.archives) {
+                            self.archives = true;
+                            $(self).triggerHandler('archives');
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Should we make this a 'changed' event, to cover other variables that change?

File: pkg/base1/cockpit.js [hunk 4]
```diff
@@ -2808,7 +2826,6 @@ function full_scope(cockpit, $, po) {
                     var message_len = message.length;
                     if (message_len === undefined) {
                         meta = message;
-                        last = null;
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: I thought all bets were off when a meta message came through. The data before/after can be discontiguous. Could you explain why this line is removed?

**Comment by stefwalter** [comment]
> Allow metrics compression across meta messages.

Why?

**Comment by stefwalter** [comment]
>  base1: Use relative timestamps when fetching metrics. So that we are protected from wrong server clocks.

I'm not sure this is correct. We should be displaying and retrieving data for the time that the server thinks it is. The point of the grid is to help data from N sources to line up into a quantifiable set of events, both for history and current data.

If the server clocks are off, so are the graphs.

But for certain single source graphs, if we want to support this, we could: have MetricsChannel() expose an offset, which is the number of intervals it has discovered that the server is different from the browser time. And the graph code could set the grid to be offset by that much.

However lets not muddy the very strong concepts of interval and (beg/end) bounds in the grid, series and fetch code.

File: pkg/shell/plot.js [hunk 5]
```diff
@@ -130,12 +130,16 @@ shell.plot = function plot(element, x_range_seconds, x_stop_seconds) {
...
-    var walk_timer;
     var result = { };
 
-    var now, x_range, x_offset, interval;
+    var interval;
+
+    var grid;
 
     function refresh() {
+        if (!grid)
```
Comment (stefwalter) [inline pkg/shell/plot.js]: Why would there ever be no grid? Shouldn't a graph always have a grid? Although the grid may have no rows (yet), and those rows may have no data (yet) ... it should always have a grid. The grid is the representation of the bounds of the graph.

File: pkg/shell/plot.js [hunk 6]
```diff
@@ -163,54 +167,53 @@ shell.plot = function plot(element, x_range_seconds, x_stop_seconds) {
...
+        if (x_range_seconds !== undefined) {
 
-        if (x_stop_seconds !== undefined)
-            x_offset = now - x_stop_seconds*1000;
-        else
-            x_offset = 0;
+            // Fill the plot with about 1000 samples, but don't sample
+            // faster than once per second.
+            //
+            // TODO - do this based on the actual size of the plot.
```
Comment (stefwalter) [inline pkg/shell/plot.js]: The interval is locked. So not sure how this would be variable. Wouldn't you just move() the g rid to match your x_range_seconds.

File: pkg/shell/plot.js [hunk 7]
```diff
@@ -163,54 +167,53 @@ shell.plot = function plot(element, x_range_seconds, x_stop_seconds) {
...
+                x_offset = (new Date().getTime()) - x_stop_seconds*1000;
+            else
+                x_offset = 0;
+
+            var beg = -Math.ceil((x_range_seconds*1000 + x_offset) / interval);
+            var end = -Math.floor(x_offset / interval);
+
+            if (grid)
+                grid.close();
+            grid = cockpit.grid(interval, beg, end);
```
Comment (stefwalter) [inline pkg/shell/plot.js]: I think you want grid.move() here.

File: pkg/shell/plot.js [hunk 8]
```diff
@@ -310,263 +287,63 @@ shell.plot = function plot(element, x_range_seconds, x_stop_seconds) {
...
+                    row[x + i] = [(grid.beg + x + i)*interval, flat_sum(metrics_row[x + i])*factor];
+            });
 
-            if (-p1 >= n_real_time_samples)
-                p1 = -n_real_time_samples + 1;
 
-            if (p1 > p2)
-                return null;
+            channel.follow();
+            channel.fetch(grid.beg, grid.end);
```
Comment (stefwalter) [inline pkg/shell/plot.js]: The grid will ask the series for data, and the series will invoke the fetch() call ... there's no need for it here.

**Comment by mvollmer** [comment]
>>  base1: Use relative timestamps when fetching metrics. So that we are protected from wrong server clocks.

> I'm not sure this is correct. 

When two clocks disagree, one of them is wrong.  I don't think there is a point in trying to make a graph that is based on wrong timestamps.

**Comment by stefwalter** [comment]
> When two clocks disagree, one of them is wrong. I don't think there is a point in trying to make a graph that is based on wrong timestamps.

Fair enough, but then the MetricsChannel should track and perform the offset, and not just make everything relative.

**Comment by stefwalter** [comment]
@mvollmer As far as the relative timestamps ... you could do that. You just need to insulate the rest of the code (series, grid) from that. So when you get the data back, you have to adjust all the timestamps in meta tags by the relative offset (which should be sorta easy to calculate).

I hope that achieves what you need.

**Comment by mvollmer** [comment]
> I hope that achieves what you need.

Yes, that is the plan.  The current code (in this PR) is buggy and incomplete.  I'll get back to this soonish.

File: pkg/base1/cockpit.js (hunk 1 above)
Comment (mvollmer) [inline pkg/base1/cockpit.js]: Yes, that's better.

File: pkg/base1/cockpit.js (hunk 2 above)
Comment (mvollmer) [inline pkg/base1/cockpit.js]: Yes, removed.

File: pkg/base1/cockpit.js (hunk 4 above)
Comment (mvollmer) [inline pkg/base1/cockpit.js]: It is up to the source whether compression is done or not across a meta message.  If it is not possible, then it will simply not be done.

File: pkg/shell/plot.js (hunk 6 above)
Comment (mvollmer) [inline pkg/shell/plot.js]: I want to use a different interval when the range changes, so that we don't sample one week of data with an interval of 1 second.

**Comment by mvollmer** [comment]
> There are some JavaScript exceptions.

Fixed.

**Comment by mvollmer** [comment]
> The cpu plot scaling is all wrong.

Fixed.

File: pkg/shell/plot.js (hunk 5 above)
Comment (mvollmer) [inline pkg/shell/plot.js]: > Why would there ever be no grid? 

That was a consequence of making x_range optional in the constructor.  I made it mandatory again.

File: pkg/shell/plot.js (hunk 7 above)
Comment (mvollmer) [inline pkg/shell/plot.js]: Yes, when the interval doesn't change.

File: pkg/shell/plot.js (hunk 8 above)
Comment (mvollmer) [inline pkg/shell/plot.js]: The cockpit.metrics did not implement the necessary hook. I have done that now, and added the 'for_walking' flag that we discussed on irc.  Scrolling now appears smoother, but after a few scrolls, things start to go wrong and both the browser and cockpit-pcp do excessive work for a long time.

**Comment by mvollmer** [comment]
This is ready for review again.

File: pkg/base1/cockpit.js [hunk 9]
```diff
@@ -2894,19 +2911,38 @@ function full_scope(cockpit, $, po) {
...
-            if (opts.source == "direct")
-                opts.source = "pcp-archive";
-            transfer(opts, drain);
+            /* Provide some fake data synchronously to the series so
+             * that it doesn't ask us again for the same range while
+             * we are still fetching it asynchronously.
+             *
+             * XXX - is this something the series should do itself?
+             */
+            self.series.input(beg, new Array(end-beg), { });
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: This code should be in the series. Agree with the comment. I think simply this bit of logic in the caller to fetch.

File: pkg/base1/cockpit.js [hunk 10]
```diff
@@ -2894,19 +2911,38 @@ function full_scope(cockpit, $, po) {
...
+
+            var timestamp = beg*interval - $.now();
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Needs whitespace around the ```*```

File: pkg/base1/cockpit.js [hunk 11]
```diff
@@ -2825,8 +2840,10 @@ function full_scope(cockpit, $, po) {
                     var message_len = message.length;
                     if (message_len === undefined) {
                         meta = message;
-                        last = null;
-                        beg = Math.floor((meta.timestamp || 0) / interval);
+                        var timestamp = 0;
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: We shouldn't define variables inside of an if statement.

File: pkg/base1/cockpit.js [hunk 12]
```diff
@@ -2894,19 +2911,38 @@ function full_scope(cockpit, $, po) {
...
         };
 
+        var following = false;
+
         self.follow = function follow() {
-            var opts = $.extend({ }, self.options);
-            delete opts.limit;
-            transfer(opts, drain);
+            if (!following) {
+                following = true;
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: It seems like following is never reset to false, when the following channel closes (for whatever reason).

File: pkg/base1/cockpit.js [hunk 13]
```diff
@@ -2977,7 +3013,7 @@ function full_scope(cockpit, $, po) {
             return low;
         }
 
-        self.load = function load(beg, end) {
+        self.load = function load(beg, end, for_walking) {
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: It seems like the arguments here are mutually exclusive. Either beg/end is relevant or for_walking is relevant? Do you think we should just "overload" the function ... so that when no arguments are specified, we load with "for_walking" behavior?

File: pkg/base1/cockpit.js [hunk 14]
```diff
@@ -3008,8 +3046,11 @@ function full_scope(cockpit, $, po) {
                 }
             }
 
+            for (i = 0; i < fetches.length; i++)
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Could you explain why this change is necessary? Why does calling fetch() interleaved with process() cause problems?

File: pkg/shell/plot.js [hunk 15]
```diff
@@ -145,8 +146,13 @@ shell.plot = function plot(element, x_range_seconds, x_stop_seconds) {
         flot.setData(data);
         var axes = flot.getAxes();
 
-        axes.xaxis.options.min = now - x_range - x_offset;
-        axes.xaxis.options.max = now - x_offset;
+        /* Walking and fetching samples are not synchronized, which
+         * means that a walk step might reveal a sample that hasn't
+         * been fetched yet.  To avoid this, we cut off one extra
```
Comment (stefwalter) [inline pkg/shell/plot.js]: I would say "To reduce flicker, we cut off one extra"

File: pkg/shell/plot.js [hunk 16]
```diff
@@ -145,8 +146,13 @@ shell.plot = function plot(element, x_range_seconds, x_stop_seconds) {
...
+         * sample at the end.
+         */
+        axes.xaxis.options.min = grid.beg*interval;
+        axes.xaxis.options.max = (grid.end-2)*interval;
```
Comment (stefwalter) [inline pkg/shell/plot.js]: Spacing around the ```*```

File: pkg/shell/plot.js [hunk 17]
```diff
@@ -162,55 +168,73 @@ shell.plot = function plot(element, x_range_seconds, x_stop_seconds) {
...
             x_offset = 0;
 
-        // Fill the plot with about 1000 samples.
-        //
-        // TODO - do this based on the actual size of the plot.
-        //
-        if (x_range_seconds !== undefined) {
-            x_range = x_range_seconds * 1000;
-            interval = Math.ceil(x_range_seconds / 1000) * 1000;
+        var beg = -Math.ceil((x_range_seconds*1000 + x_offset) / interval);
```
Comment (stefwalter) [inline pkg/shell/plot.js]: Whitespace

File: pkg/shell/plot.js [hunk 18]
```diff
@@ -310,267 +308,72 @@ shell.plot = function plot(element, x_range_seconds, x_stop_seconds) {
...
+                               });
         }
 
-        function build_metric(n) {
-            return { name: n, units: desc.units, derive: desc.derive };
+        function flat_sum(val) {
+            if (!val)
+                return 0;
+            if (val.length !== undefined) {
+                var i, sum = 0;
```
Comment (stefwalter) [inline pkg/shell/plot.js]: We shouldn't declare variables inside of if statement.

**Comment by stefwalter** [comment]
Lots of really good fixes here. Just a few more issues and questions.

**Comment by stefwalter** [comment]
If you look in the tests, you can see hundreds of lines like this:

```

> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
.
----------------------------------------------------------------------
Ran 1 test in 25.845s

OK
.
----------------------------------------------------------------------
Ran 1 test in 25.257s

OK
> Authentication failed
> Authentication failed
> Permission denied
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
..
----------------------------------------------------------------------
Ran 2 tests in 71.457s

OK
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
.
----------------------------------------------------------------------
Ran 1 test in 54.706s

OK
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
.
----------------------------------------------------------------------
Ran 1 test in 80.987s

OK
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
Write failed: Broken pipe
> transport closed: disconnected
> transport closed, dropped message:  
{"command":"open","channel":"1","payload":"null","host":"localhost"}
> transport closed, dropped message:  
{"command":"kill","group":"localhost/shell/shell"}
> transport closed, dropped message:  
{"command":"logout","disconnect":true}
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
.
----------------------------------------------------------------------
Ran 1 test in 61.248s

OK
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
.> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
.
----------------------------------------------------------------------
Ran 2 tests in 59.971s

OK
> fatal: 0 error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> transport closed: disconnected
> fatal: 0 error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> health check failed
> transport closed: timeout
> transport closed, dropped message:  
{"command":"open","channel":"1","payload":"null","host":"localhost"}
> transport closed, dropped message:  
{"command":"kill","group":"localhost/shell/shell"}
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
..
----------------------------------------------------------------------
Ran 2 tests in 72.371s

OK
.Created symlink from /etc/systemd/system/default.target.wants/test.service to /usr/lib/systemd/system/test.service.
.
----------------------------------------------------------------------
Ran 2 tests in 70.698s

OK
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> Client is already subscribed.
> transport closed: disconnected
> transport closed, dropped message:  
{"command":"open","channel":"1","payload":"null","host":"localhost"}
> transport closed, dropped message:  
{"command":"kill","group":"localhost/system/init"}
> transport closed, dropped message:  
{"command":"logout","disconnect":true}
.> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
> metrics channel failed: internal-error
.
----------------------------------------------------------------------
Ran 2 tests in 50.231s

OK
> metrics channel failed: not-found
> metrics channel failed: not-found
.
----------------------------------------------------------------------
Ran 1 test in 29.257s

OK
...
----------------------------------------------------------------------
Ran 3 tests in 104.417s

OK
> metrics channel failed: not-found
> metrics channel failed: not-found
.
----------------------------------------------------------------------
Ran 1 test in 28.336s

OK
> metrics channel failed: internal-error
> metrics channel failed: internal-error
.
----------------------------------------------------------------------
Ran 1 test in 40.589s

OK
> metrics channel failed: internal-error
> metrics channel failed: internal-error
.
----------------------------------------------------------------------
Ran 1 test in 28.463s

OK
> metrics channel failed: not-found
> metrics channel failed: not-found
.
----------------------------------------------------------------------
Ran 1 test in 43.960s

OK
> metrics channel failed: not-found
> metrics channel failed: not-found
.
----------------------------------------------------------------------
Ran 1 test in 46.697s

OK
> metrics channel failed: not-found
> metrics channel failed: not-found
.
----------------------------------------------------------------------
Ran 1 test in 37.634s

OK
> metrics channel failed: internal-error
> metrics channel failed: internal-error
.
----------------------------------------------------------------------
Ran 1 test in 59.972s

OK
> metrics channel failed: not-found
> metrics channel failed: not-found
Warning: The resulting partition is not properly aligned for best performance.
.
----------------------------------------------------------------------
Ran 1 test in 26.325s

OK
Password for admin@COCKPIT.LAN: 
.> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> Failed to join domain: cockpit.lan: Password is incorrect
.
----------------------------------------------------------------------
Ran 2 tests in 276.532s

OK
> metrics channel failed: not-found
> metrics channel failed: not-found
.> metrics channel failed: not-found
> metrics channel failed: not-found
.
----------------------------------------------------------------------
Ran 2 tests in 70.853s

OK
> metrics channel failed: internal-error
> metrics channel failed: internal-error
.
----------------------------------------------------------------------
Ran 1 test in 31.195s

OK
> metrics channel failed: not-found
> metrics channel failed: not-found
.
----------------------------------------------------------------------
Ran 1 test in 27.680s

OK
> metrics channel failed: not-found
> metrics channel failed: not-found
.
----------------------------------------------------------------------
Ran 1 test in 28.011s

OK
.
----------------------------------------------------------------------
Ran 1 test in 22.104s

OK
.
----------------------------------------------------------------------
Ran 1 test in 23.965s

OK
> transport closed, dropped message:  
{"command":"kill","group":"localhost/users/local"}
.
----------------------------------------------------------------------
Ran 1 test in 38.958s

OK
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> transport closed, dropped message:  
{"command":"kill","group":"localhost/shell/shell"}
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> transport closed: terminated
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
.
----------------------------------------------------------------------
Ran 1 test in 42.573s

OK
> couldn't set machine color: Not permitted to perform this action.
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> Unexpected error Not permitted to perform this action.
> Unexpected error Not permitted to perform this action.
> Unexpected error Not permitted to perform this action.
> transport closed, dropped message:  
{"command":"kill","group":"localhost/users/local"}
> transport closed, dropped message:  
{"command":"kill","group":"localhost/shell/shell"}
..
----------------------------------------------------------------------
Ran 2 tests in 59.620s

OK
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> transport closed, dropped message:  
{"problem":"cancelled","command":"close","channel":"1:1!14"}
> transport closed, dropped message:  
{"command":"close","channel":"1:1!17"}
> transport closed, dropped message:  
{"command":"close","channel":"1:1!18"}
> transport closed, dropped message:  
{"command":"close","channel":"1:1!22"}
> transport closed, dropped message:  
{"command":"kill","group":"localhost/shell/shell"}
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> transport closed, dropped message:  
{"command":"kill","group":"10.111.111.16/shell/shell"}
> transport closed, dropped message:  
{"command":"kill","group":"localhost/shell/shell"}
.
----------------------------------------------------------------------
Ran 1 test in 68.579s

OK
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-found
> metrics channel failed: not-supported
> metrics channel failed: not-supported
> metrics channel failed: not-supported
> metrics channel failed: not-supported
> metrics channel failed: not-supported
> metrics channel failed: not-supported
> metrics channel failed: not-supported
> metrics channel failed: not-supported
> metrics channel failed: not-supported
> metrics channel failed: not-supported
> metrics channel failed: not-supported
> metrics channel failed: not-supported
.
----------------------------------------------------------------------
Ran 1 test in 75.504s
```

File: pkg/base1/cockpit.js (hunk 9 above)
Comment (mvollmer) [inline pkg/base1/cockpit.js]: Done in a FIXUP.

File: pkg/base1/cockpit.js (hunk 14 above)
Comment (mvollmer) [inline pkg/base1/cockpit.js]: Done in the FIXUP.

File: pkg/shell/plot.js (hunk 15 above)
Comment (mvollmer) [inline pkg/shell/plot.js]: Yep.

File: pkg/shell/plot.js (hunk 18 above)
Comment (mvollmer) [inline pkg/shell/plot.js]: Done.

File: pkg/shell/plot.js (hunk 17 above)
Comment (mvollmer) [inline pkg/shell/plot.js]: Done.

File: pkg/shell/plot.js (hunk 16 above)
Comment (mvollmer) [inline pkg/shell/plot.js]: Done.

File: pkg/base1/cockpit.js (hunk 11 above)
Comment (mvollmer) [inline pkg/base1/cockpit.js]: Done.

File: pkg/base1/cockpit.js (hunk 10 above)
Comment (mvollmer) [inline pkg/base1/cockpit.js]: Done.

File: pkg/base1/cockpit.js (hunk 12 above)
Comment (mvollmer) [inline pkg/base1/cockpit.js]: Good catch, fixed.

File: pkg/base1/cockpit.js (hunk 13 above)
Comment (mvollmer) [inline pkg/base1/cockpit.js]: Hmm, the fact that beg, end is always present means that people can simply ignore for_walking and the right thing will happen.  Your cadvisor polling code will not need any change, for example.

**Comment by mvollmer** [comment]
> If you look in the tests, you can see hundreds of lines like this:

Most of these come from trying to read the pcp archives while none are present.  In addition, we would return erroneously "internal-error" when the directory for pmlogger exists but is empty.

I don't know yet about the "not-supported" errors.  We should probably just ignore them, too, for archives.

**Comment by mvollmer** [comment]
> I don't know yet about the "not-supported" errors. We should probably just ignore them, too, for archives.

I did that.

**Comment by mvollmer** [comment]
Rebased and squashed.

**Comment by mvollmer** [comment]
Right, unit tests are still failing.

**Comment by mvollmer** [comment]
The integration test failures seem genuine.

**Comment by stefwalter** [comment]
This failure is being fixed in #2774:

```
Page error: TypeError: 'undefined' is not an object (evaluating 'p.LoadError[1]')
```

**Comment by stefwalter** [comment]
Yay. Merged.
